src/cfmms.jl

```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
+    ticks #This is the tick mapping sorted by price
+    function UniV3(R,γ,idx,current_price,current_tick_index,ticks)
+        γ_T, idx_uint, T = two_coin_check_cast(R, γ, idx)
+        return new{T}(
+            MVector{2,T}(R),
+            γ_T,
+            MVector{2,UInt}(idx_uint),
+            current_price,
+            current_tick_index,
+            ticks
+        )
+    end
+end
+
+## See univ3 whitepaper
+function virtual_reserves(P,L)
+    sP = sqrt(P)
+    x = L/sP
+    y = L*sP
+    return([x, y])
+end
+
+function find_arb!(Δ::VT, Λ::VT, cfmm::UniV3{T}, v::VT) where {T, VT<:AbstractVector{T}} 
+    current_price, current_tick_index, γ, ticks = cfmm.current_price, cfmm.current_tick_index, cfmm.γ, cfmm.ticks
+    Δ[1] = 0
+    Δ[2] = 0
+
+    Λ[1] = 0
+    Λ[2] = 0
+
+    #target_price = max(min(v[1]/v[2]/γ,2.0^64),1/2.0^64) #including out of bounds check here
+    target_price = v[1]/v[2]
+    if (target_price*γ < current_price) && (current_price < target_price * 1/(γ))
+        return nothing
+    end
+    target_price = target_price/γ
+    if target_price >= current_price #iterate forwards in the tick mapping
+        i = 1
+        while true
+            next_tick_price = ticks[current_tick_index + i]["price"]
+            if next_tick_price >= target_price ## so now we know that current_price <= target_price < next_tick_price
+                #print(max(current_price,ticks[current_tick_index + i - 1]["price"]))
+                R = virtual_reserves(max(current_price,ticks[current_tick_index + i - 1]["price"]),ticks[current_tick_index + i - 1]["liquidity"])
+                #print(R[2]/R[1])
+                k = R[1]*R[2]
+
+                Δ[1] += prod_arb_δ(1/target_price, R[1], k, 1)
+                Δ[2] += prod_arb_δ(target_price, R[2], k, 1)
+
+                Λ[1] += prod_arb_λ(target_price, R[1], k, 1)
+                Λ[2] += prod_arb_λ(1/target_price, R[2], k, 1)
+
+                break
+
+            elseif next_tick_price < target_price ## so now we know that current_price <= next_tick_price <= target_price
+                R = virtual_reserves(max(current_price,ticks[current_tick_index + i - 1]["price"]),ticks[current_tick_index + i - 1]["liquidity"])
+                k = R[1]*R[2]
+
+                Δ[1] += prod_arb_δ(1/next_tick_price, R[1], k, 1)
+                Δ[2] += prod_arb_δ(next_tick_price, R[2], k, 1)
+
+                Λ[1] += prod_arb_λ(next_tick_price, R[1], k, 1)
+                Λ[2] += prod_arb_λ(1/next_tick_price, R[2], k, 1)
+            end
+            i += 1
+        end
+
+    elseif target_price < current_price #iterate backwards in the tick mapping
+        i = 1
+        while true
+            prev_tick_price = ticks[current_tick_index - i + 1]["price"]
+            if prev_tick_price <= target_price ## so now we know that prev_tick_price < target_price <=  current_price
+
+                print(ticks[current_tick_index - i + 2]["price"])
```

Comment:
remove print statements (or add `verbose` kwarg so all prints are changed to `verbose && print(...)`)